pallets/protocol-fee/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+    trait Store for Module<T: Trait> as ProtocolFee {
+        /// The mapping of operation names to the base fees of those operations.
+        pub BaseFees get(base_fees) config(): map ProtocolOp => BalanceOf<T>;
+        /// The fee coefficient as a positive rational (numerator, denominator).
+        pub Coefficient get(coefficient) config() build(|config: &GenesisConfig<T>| {
+            if config.coefficient.1 == 0 {
+                PosRatio(1, 1)
+            } else {
+                config.coefficient
+            }
+        }): PosRatio;
+    }
+}
+
+decl_event! {
+    pub enum Event<T> where Balance = BalanceOf<T> {
+        /// The protocol fee of an operation.
+        Fee(Balance),
+        /// The fee coefficient.
+        Coefficient(PosRatio),
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Changes the fee coefficient for the root origin.
+        pub fn change_coefficient(origin, coefficient: PosRatio) -> DispatchResult {
+            ensure_root(origin)?;
+            <Coefficient>::put(coefficient);
+            Ok(())
+        }
+
+        /// Changes the a base fee for the root origin.
+        pub fn change_base_fee(origin, op: ProtocolOp, base_fee: BalanceOf<T>) ->
+            DispatchResult
+        {
+            ensure_root(origin)?;
+            <BaseFees<T>>::insert(op, base_fee);
+            Ok(())
+        }
+
+        /// Emits an event with the fee of the operation.
+        pub fn get_fee(_origin, op: ProtocolOp) -> DispatchResult {
+            let fee = Self::compute_fee(op)?;
+            Self::deposit_event(RawEvent::Fee(fee));
+            Ok(())
+        }
+
+        /// Emits an event with the fee coefficient.
+        pub fn get_coefficient(_origin) -> DispatchResult {
+            Self::deposit_event(RawEvent::Coefficient(Self::coefficient()));
+            Ok(())
+        }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    /// Computes the fee of the operation as `(base_fee * coefficient.0) / coefficient.1`.
+    pub fn compute_fee(op: ProtocolOp) -> ComputeFeeResult<T> {
+        let coefficient = Self::coefficient();
+        let (numerator, denominator) = (coefficient.0, coefficient.1);
+        if let Some(fee) = Self::base_fees(op)
+            .saturating_mul(<BalanceOf<T>>::from(numerator))
+            .checked_div(&<BalanceOf<T>>::from(denominator))
+        {
+            Ok(fee)
+        } else {
+            Err(Error::<T>::ComputeFee.into())
+        }
+    }
+
+    /// Computes the fee of the operation and charges it to the given signatory.
+    pub fn charge_fee(signatory: &Signatory, op: ProtocolOp) -> DispatchResult {
+        let fee = Self::compute_fee(op)?;
+        if fee.is_zero() {
+            return Ok(());
+        }
+        let imbalance = Self::withdraw_fee(signatory, fee)?;
+        T::OnProtocolFeePayment::on_unbalanced(imbalance);
```

Comment:
It is better to add a code comment about where the balance goes.. or the perspective of `on_unbalanced`

Reply:
This is already stated in the doc comment of this method.

Reply:
Also the recipient of the fee depends on the implementation of `OnProtocolFeePayment`, and that's in `pallets/runtime/src/runtime.rs`.

Reply:
Comments added.